### PR TITLE
[ankurvdev-embedresource] adding new port

### PR DESCRIPTION
**Pull request by ankurvdev**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by dg0yt** [comment]
This needs a host tool when consomed. We must ensure that everything is always used from the vcpkg installation, not loaded via FetchContent. It might help to have a CI test port.
(BTW it is not the first resource compiler port.)

**Comment by ankurvdev** [comment]
> This needs a host tool when consomed. We must ensure that everything is always used from the vcpkg installation, not loaded via FetchContent. It might help to have a CI test port.

I've got vcpkg tests in the source repo already.
Did you mean in the vcpkg repo ?
Is there a pattern i can follow for creating a ci test port ?

> (BTW it is not the first resource compiler port.)

All I could find was cmakerc.
Can you point me to a few others if they exist ?

**Comment by dg0yt** [comment]
I consider the resource compiler topic sensitive to cross-build issues. That's why I ask for test port, given the young project. Disclaimer: I'm only a community member.

> Did you mean tests in the vcpkg repo ? Is there a pattern i can follow for creating a ci test port ?

This one was just merged: 
https://github.com/microsoft/vcpkg/tree/master/scripts/test_ports/vcpkg-ci-soci
It is a simple portfile to configure and build a simple cmake project. This isn't only run in CI but can be used easily in manual tests with `--overlay-ports=scripts/test-ports`.
(Adding the test port immediately uncovered problems with dllexport declarations so creating it was immediately useful.)

> I've got vcpkg tests in the source repo already.[](https://github.com/ankurvdev/embedresource/blob/main/ci/sample/CMakeLists.txt)

The idea is to have a CMake project which is *not tightly coupled* with vcpkg (in the form of [ci/sample/CMakeLists.txt](https://github.com/ankurvdev/embedresource/blob/acbb1736f46da70bb06d33c53da5bd2dcdc2fc85/ci/sample/CMakeLists.txt#L3-L15)), but uses the minimal CMake syntax *without any assumptions about vcpkg*, such as [the soci test project](https://github.com/microsoft/vcpkg/blob/master/scripts/test_ports/vcpkg-ci-soci/project/CMakeLists.txt). Basically the [example from the README](https://github.com/ankurvdev/embedresource/blob/acbb1736f46da70bb06d33c53da5bd2dcdc2fc85/README.md?plain=1#L27-L31), decorated with a project header.
This would run in vcpkg CI for `arm-neon-android` (32 bit) on `x64-linux` (64 bit) host - is this covered in your CI?

> > (BTW it is not the first resource compiler port.)
> 
> All I could find was cmakerc. Can you point me to a few others if they exist ?

Maybe `cmakerec` is the only lean port, but there are big fishes which bring own tools, at least Qt. (The qt5 ports do not support cross builds although upstream does.)

**Comment by ankurvdev** [comment]
Thanks for the pointers. Yeah I agree these sorts of cross-compile projects are extremely useful for any form of codegen tools

The CI project in source repo does cover cross-compilaton for the following triplets [build](https://dev.azure.com/ankurv/embedresource/_build/results?buildId=695&view=results)
* arm*-android
* wasm32-emscripten (this one is especially tricky)
* uwp
(and mingw too)

Basically, The ci sample project is used to test both vcpkg-mode and add_subdirectory-mode (no cross-compile) which are listed on the README as integration choices.
(probably why it might seem a little less lean)
But when operating in vcpkg mode it relies on `find_package` which is the only thing needed.

That said, if you think it'll be prudent, 
I can definitely follow the pattern you listed and wouldn't mind creating a leaner "vcpkg-only" test-port in the folder.

**Comment by ankurvdev** [comment]
@dg0yt 
I added a CI test and looks like I was mistaken and apparently find_program isn't working as I expected it.
My guess is I need to tweak [these lines](https://github.com/ankurvdev/embedresource/blob/acbb1736f46da70bb06d33c53da5bd2dcdc2fc85/cmake/EmbedResourceTargets.cmake#L4C1-L7C8)
but I'm not sure what the right pattern here is

Could you advise ?

**Comment by dg0yt** [comment]
That `find_program` pattern is unusual, but the vcpkg toolchain wouldn't do the right thing anyways.

I would add a post-install fixup in the vcpkg portfile, based on the known vcpkg install layout. Something like
~~~cmake
file(READ "${CURRENT_PACKAGES_DIR}/share/embedresource/EmbedResourceConfig.cmake" config)
file(WRITE "${CURRENT_PACKAGES_DIR}/share/embedresource/EmbedResourceConfig.cmake" "
find_program(EMBEDRESOURCE_EXECUTABLE embedresource PATHS "\${CMAKE_CURRENT_LIST_DIR}/../../../${HOST_TRIPLET}/tools/embedresource" REQUIRED)
\${config}")
~~~

File: scripts/test_ports/vcpkg-ci-embedresource/portfile.cmake [hunk 1]
```diff
@@ -1,4 +1,8 @@
 set(VCPKG_POLICY_EMPTY_PACKAGE enabled)
 
-vcpkg_cmake_configure(SOURCE_PATH "${CURRENT_PORT_DIR}/project")
+vcpkg_cmake_configure(
+  SOURCE_PATH "${CURRENT_PORT_DIR}/project"
+  OPTIONS
+    -DVCPKG_HOST_TRIPLET=${_HOST_TRIPLET}
```
Comment (dg0yt) [inline scripts/test_ports/vcpkg-ci-embedresource/portfile.cmake:6]: I know why you do it... but you can't rely on the users doing it when even `vcpkg_cmake_configure` doesn't do it.
(Related: https://github.com/microsoft/vcpkg/pull/25529.)

**Comment by ankurvdev** [comment]
Thanks for the suggestion. That works !

File: scripts/test_ports/vcpkg-ci-embedresource/portfile.cmake (hunk 1 above)
Comment (ankurvdev) [reply scripts/test_ports/vcpkg-ci-embedresource/portfile.cmake:6]: Yeah I agree . 
Still looking for a good alternative

**Comment by ankurvdev** [comment]
> ```cmake
> file(READ "${CURRENT_PACKAGES_DIR}/share/embedresource/EmbedResourceConfig.cmake" config)
> file(WRITE "${CURRENT_PACKAGES_DIR}/share/embedresource/EmbedResourceConfig.cmake" "
> find_program(EMBEDRESOURCE_EXECUTABLE embedresource PATHS "\${CMAKE_CURRENT_LIST_DIR}/../../../${HOST_TRIPLET}/tools/embedresource" REQUIRED)
> \${config}")

That didnt quite work. 

HOST_TRIPLET wasnt defined

**Comment by dg0yt** [comment]
> That didnt quite work.
> 
> HOST_TRIPLET wasnt defined

Be careful about the variables and escaping. `${HOST_TRIPLET}` is to be evaluated in the portfile where it is defined. Same for `${config}` but I accidentally added a backslash :facepalm:. `${CMAKE_CURRENT_LIST_DIR}` is really meant to be evaluated when it is is used, so this needs (and has) the backslash.

**Comment by ankurvdev** [comment]
Thanks 
Apparently a NO_DEFAULT_PATH was required as well for android to work correctly.

File: ports/embedresource/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# sha512sum can be obtained using
+# wget -O - -q  https://github.com/ankurdev/embedresource/archive/b6429f8b92947273a5e66d5f10210b960616a89d.tar.gz | sha512sum 
+set(commitId acbb1736f46da70bb06d33c53da5bd2dcdc2fc85)
+set(sha512 06c1692f89ed37db8978afb6d878b1882d0e357bc31f5fbdb6a8b6f07041b152d12379ece617a851483d1582d0c2af5443877117f3da9cc0ca6919dfdcfd81fe)
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO ankurvdev/embedresource
+    REF ${commitId}
+    SHA512 ${sha512}
+    HEAD_REF master)
```
Comment (Cheney-W) [inline ports/embedresource/portfile.cmake:10]:
```suggestion
vcpkg_from_github(
    OUT_SOURCE_PATH SOURCE_PATH
    REPO ankurvdev/embedresource
    REF "v${VERSION}"
    SHA512 96d2208fd5d654dad5662968296fa363cea0a935fec8474b780717c9303d2dd763833370bcdf02d6d63e264368b0955fa1f13c6e55685280df5fdaf9e72b8c9f
    HEAD_REF main)
```

File: ports/embedresource/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
+)
+
+vcpkg_cmake_install()
+vcpkg_copy_pdbs()
+
+# Handle copyright
+file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```
Comment (Cheney-W) [inline ports/embedresource/portfile.cmake:24]:
```suggestion
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
```

File: ports/embedresource/vcpkg.json [hunk 4]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "embedresource",
+  "version": "0.0.10",
+  "description": "Cross Platform Resource Embedding",
```
Comment (Cheney-W) [inline ports/embedresource/vcpkg.json:4]:
```suggestion
  "description": "Cross Platform Resource Embedding",
  "homepage": "https://github.com/ankurvdev/embedresource",
```

File: ports/embedresource/vcpkg.json [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+      "name": "embedresource",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
```
Comment (Cheney-W) [inline ports/embedresource/vcpkg.json:17]:
```suggestion
```

File: ports/embedresource/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO ankurvdev/embedresource
+    REF "v${VERSION}"
+    SHA512 96d2208fd5d654dad5662968296fa363cea0a935fec8474b780717c9303d2dd763833370bcdf02d6d63e264368b0955fa1f13c6e55685280df5fdaf9e72b8c9f
+    HEAD_REF main)
+
+if(NOT TARGET_TRIPLET STREQUAL HOST_TRIPLET)
+    vcpkg_add_to_path(PREPEND "${CURRENT_HOST_INSTALLED_DIR}/tools")
+endif()
+
```
Comment (dg0yt) [inline ports/embedresource/portfile.cmake:11]:
```suggestion
```

File: ports/embedresource/portfile.cmake [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    SOURCE_PATH ${SOURCE_PATH}
+)
+
+vcpkg_cmake_install()
+vcpkg_copy_pdbs()
+
+# Handle copyright
+vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")
+file(RENAME "${CURRENT_PACKAGES_DIR}/bin" "${CURRENT_PACKAGES_DIR}/tools")
```
Comment (dg0yt) [inline ports/embedresource/portfile.cmake:22]: This is neither the usual pattern nor the right destination. Use `vcpkg_copy_tools(TOOL_NAMES ... AUTO_CLEAN)`.

File: ports/embedresource/portfile.cmake [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+file(READ "${CURRENT_PACKAGES_DIR}/share/embedresource/EmbedResourceConfig.cmake" config_contents)
+file(WRITE "${CURRENT_PACKAGES_DIR}/share/embedresource/EmbedResourceConfig.cmake"
+"find_program(
+    EMBEDRESOURCE_EXECUTABLE embedresource
+    PATHS
+        \"\${CMAKE_CURRENT_LIST_DIR}/../../../${HOST_TRIPLET}/tools\"
```
Comment (dg0yt) [inline ports/embedresource/portfile.cmake:29]:
```suggestion
        \"\${CMAKE_CURRENT_LIST_DIR}/../../../${HOST_TRIPLET}/tools/${PORT}\"
```
(after using `vcpkg_copy_tools`).

**Review by dg0yt — COMMENTED** [review]
I have another set of comments. It should be the last set from my side (community).

**Review by Cheney-W — DISMISSED** [review]

**Comment by ankurvdev** [comment]
Can you suggest a prefix ?

**Comment by JavierMatosD** [comment]
> Can you suggest a prefix ?

ankurvdev-embedresource?

**Pull request merged**